csharp/src/Ice/Connection.cs

```diff
@@ -1134,7 +1128,8 @@ private void Dispatch(IStartCallback? startCB, Queue<OutgoingMessage>? sentCBs,
                     {
                         Debug.Assert(m.OutAsync != null);
                         var outAsync = (OutgoingAsync)m.OutAsync;
-                        if (outAsync.Response())
+                        Debug.Assert(m.ResponseFrame != null);
+                        if (outAsync.Response(m.ResponseFrame))
```

Comment:
It would be better to provide the Data to Response just like with provide the Data to InvokeAsync for requests... this way we don't have to keep around the ResponseFrame in OutgoingMessage and ParseMessage doesn't need to construct it. It's constructed only when needed from the dispatch thread by OutgoingAsync.Response.